### PR TITLE
Plumb dynamic shape support through for Vulkan and VMVX

### PR DESCRIPTION
**Pull request by antiagainst**
This commit plumbs dynamic shape support through for both Vulkan and
VMVX. They rely on 1-D MemRef and running `FlattenMemRefSubspanPass`
in advance, instead of MemRef descriptors.

In order to enable dynamic shape support, we need to carry the
SSA values for dynamic dimensions down the CodeGen pipeline so that
we can linearize the index calculation in `FlattenMemRefSubspanPass`.
We have such information tightly associated with various ops at
the Flow level, but when outlining executables and materializing
HAL interface, the association is broken down. Instead, `tie_shape`
ops are used to carry such information. It's structurally difficult
to maintain and convert.

So this commit changes the `hal.interface.binding.subspan` to carry
the dynamic dimension SSA values by itself, like many other ops
in Flow/HAL. It's a natural change that simplifies lots of analysis
and transformation. For example, we don't need to maintain the two
step conversion at CPU side (first generating an undefined MemRef
descriptor when handling the `subspan` op, and then filling its
content when handling the `tie_shape` op). It also makes the
intervals of HAL more akin to Flow on this front.

Other changes are mostly natural based on that:

* `MaterializeInterfaces` picks up the information from `tie_shape`
  ops and attaches them to `subspan` ops.
* `FlattenBindingSubspan` reads the dynamic dimensions to perform
  index linearization.
* `ConvertToLLVM` now generates the full MemRef descriptor from
  `subspan` ops.
* A new pass is added to fold `memref.dim`/`tensor.dim` ops over
  shape carrying ops.

This puts IREE CodeGen dynamic shape support for Vulkan/VMvX in
a very nice state:

Because we run `FoldSubViewOpsPass` in advance, there won't be
intermediate MemRefs (coming from `subview` ops). So load/stores
directly take in HAL `subspan` ops. By definition in IREE we have
tightly packed buffers so all MemRefs coming from subspans should
have strides of the total element count in inner dimensions. So symbolic
strides in subspan ops' AffineMaps correspond to SSA values for
dimension sizes (or their products). Offsets are attached to subspan
ops as SSA values, but then they are "transferred" to load/store ops
during memref flattening, by being part of the index linearization
calculation.

### CODE REVIEW

**Comment by antiagainst** [comment]
It needs https://github.com/llvm/llvm-project/commit/a5621e26dbc90be3d5af164978f6fcc398f0a1d1 to be pulled in via integration. So right now the tests will fail.

File: iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp [hunk 1]
```diff
@@ -232,6 +232,7 @@ class HALDispatchABI {
   // |baseOffset| can optionally adjust the base byte offset of the buffer.
   MemRefDescriptor loadBinding(Location loc, int64_t ordinal,
                                Value baseOffsetValue, MemRefType memRefType,
+                               ValueRange  dynamicDims,
```
Comment (benvanik) [inline iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp:235]:
```suggestion
                               ValueRange dynamicDims,
```

File: iree/compiler/Dialect/HAL/Transforms/MaterializeInterfaces.cpp [hunk 2]
```diff
@@ -596,15 +597,43 @@ FuncOp InterfaceBuilder::buildRegionFuncOp() {
...
         auto bindingSymRefAttr = entryBuilder.getSymbolRefAttr(
             interfaceOp.sym_name(),
             {entryBuilder.getSymbolRefAttr(value.bindingOp)});
+
+        SmallVector<Value, 4> dynamicDims;
+        auto blockArgType = blockArg.getType().cast<Flow::DispatchTensorType>();
+        OpBuilder::InsertionGuard guard(entryBuilder);
+
+        if (!blockArgType.hasStaticShape()) {
+          // We can expect its first user to be a tie_shape op to associate
```
Comment (benvanik) [inline iree/compiler/Dialect/HAL/Transforms/MaterializeInterfaces.cpp:610]: It's been on my list to fix that for awhile - and this PR is a good start! If we update the flow op to carry the shape (either via a flow.func or arg attributes on the builtin.func referencing/etc) then we can (I think) remove all these tie_shapes!

**Review by benvanik — APPROVED** [review]
Didn't look too closely at the codegen side, but the rest looks great!

**Comment by MaheshRavishankar** [comment]
Thanks! A fairly big change. Ill take a look soon.

**Comment by stellaraccident** [comment]
Haven't reviewed but just saw this go by. And big +1. Thanks, Lei, for a grungy but necessary chunk of work!

File: iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp [hunk 3]
```diff
@@ -168,39 +166,96 @@ struct FlattenBindingSubspan final
...
-  for (auto pair : llvm::zip(indices, strides)) {
-    Value stride = builder.create<ConstantIndexOp>(loc, std::get<1>(pair));
+  int64_t offset;
+  SmallVector<int64_t, 4> strides;
+  strides.reserve(rank);
+
+  // First try to get the strides from the MemRef type itself. This applies to
+  // cases where we have static shapes and only the leading dimension is
+  // dynamic.
+  if (succeeded(getStridesAndOffset(sourceType, strides, offset)) &&
```
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp:229]: Could just use `getStridedLayoutMap` (https://github.com/llvm/llvm-project/blob/3419e85b15e35e4b5b4d0cd8153fdbe0958e80c7/mlir/include/mlir/IR/BuiltinTypes.h#L407)

File: iree/compiler/Codegen/Common/ShapeToLLVMConversion.cpp [hunk 4]
```diff
@@ -1,104 +0,0 @@
-// Copyright 2021 The IREE Authors
```
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/Common/ShapeToLLVMConversion.cpp:1]: Wondering why this file was deleted. Does this completely drop shape dialect and therefore these patterns are no more needed?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Just a few questions, but overall looks fine to me.

File: iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp (hunk 3 above)
Comment (antiagainst) [reply iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp]: Cool! Didn't know about that utility function before.

File: iree/compiler/Codegen/Common/ShapeToLLVMConversion.cpp (hunk 4 above)
Comment (antiagainst) [reply iree/compiler/Codegen/Common/ShapeToLLVMConversion.cpp:1]: Yup. `MaterializeInterfaces` controls what the backend and HAL executable translation pipelines see. `tie_shape` ops are now removed in `MaterializeInterfaces` so backends shouldn't see them at all. Removing this doesn't break anything . That's a sign of that. (I might miss something somewhere though given those shape stuff is everywhere. But I believe the leftover should be minor issues and we can clean up when they pop up. Removing these patterns also makes those reveal easily, as if they fall through compilation will fail.)

File: iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp [hunk 5]
```diff
@@ -168,39 +166,96 @@ struct FlattenBindingSubspan final
...
+      dynamicDim = rewriter.create<ConstantIndexOp>(subspanOp.getLoc(),
+                                                    oldType.getNumElements());
+    } else {
+      ArrayRef<int64_t> oldShape = oldType.getShape();
+      Location loc = subspanOp.getLoc();
+
+      AffineExpr sym0, sym1;
+      bindSymbols(rewriter.getContext(), sym0, sym1);
+      MLIRContext *context = rewriter.getContext();
+      auto mulMap = AffineMap::get(0, 2, {sym0 * sym1}, context);
```
[the remark this answers is not in the document]
Comment (antiagainst) [reply iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp:183]: Yes `-canonicalize` combines them (as shown by tests) so I won't worry about doing that manually here.

File: iree/compiler/Codegen/Common/CleanupBufferAllocViewPass.cpp [hunk 6]
```diff
@@ -51,6 +52,12 @@ struct FoldReshapeIntoInterfaceTensorLoad : OpRewritePattern<TensorReshapeOp> {
 
   LogicalResult matchAndRewrite(TensorReshapeOp reshapeOp,
                                 PatternRewriter &rewriter) const override {
+    auto reshapeSrcType = reshapeOp.src().getType().template cast<ShapedType>();
+    auto reshapeDstType = reshapeOp.getType().template cast<ShapedType>();
+    if (!reshapeSrcType.hasStaticShape() || !reshapeDstType.hasStaticShape()) {
```
[the remark this answers is not in the document]
Comment (antiagainst) [reply iree/compiler/Codegen/Common/CleanupBufferAllocViewPass.cpp:58]: Oh, it's just NYI. So I put a TODO here. Will come to this later when it's needed.

File: iree/compiler/Dialect/HAL/Transforms/MaterializeInterfaces.cpp (hunk 2 above)
Comment (antiagainst) [reply iree/compiler/Dialect/HAL/Transforms/MaterializeInterfaces.cpp:601]: +1. This basically removes `tie_shape` ops from various backends given that `MaterializeInterfaces` is the gatekeeper for all of them. :)

File: iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp:183]: It would be simpler code to just do that instead of creating all the affine apply ops if they are going to be canonicalized anyway

File: iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp [hunk 7]
```diff
@@ -168,39 +166,92 @@ struct FlattenBindingSubspan final
...
+  SmallVector<int64_t, 4> strides;
+  strides.reserve(rank);
+
+  // First try to get the strides from the MemRef type itself. This applies to
+  // cases where we have static shapes and only the leading dimension is
+  // dynamic.
+  if (AffineMap linearLayoutMap = getStridedLinearLayoutMap(sourceType)) {
+    // Dynamic strides/offset will create symbols. There should be none for the
+    // static case.
+    if (linearLayoutMap.getNumSymbols() == 0) {
```
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp:223]: I would expect that the dynamic shape will create dynamic strides and offsets. I am missing why it is done this way.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Still have a few questions. Sorry, just trying to understand what of this is IREE specific, and what is generalizable.

(also, please do not resolve conversations, I find it hard to follow up on previous comments)

File: iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (antiagainst) [reply iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp]: That's fine with me. Changed.

File: iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp (hunk 7 above)
Comment (antiagainst) [reply iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp:222]: This is for the static case. Dynamic dimensions are handled later.

Comment (MaheshRavishankar) [reply iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp:222]: I am still not sure I follow. The case below is only handled for `ShapeCarryingInterface` ops that are only ops in IREE. We would probably need to handle the case where the symbols are dynamic "soonish". This probably the ABI side of things, but there might be memrefs within the dispatch region that also have dynamic stides and offsets. Probably with subviews. I guess this pass is run after subview folding, so maybe it doesnt matter.
Stamping it, but still not very clear what is handled and what is not.

**Review by MaheshRavishankar — APPROVED** [review]

File: iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp (hunk 7 above)
Comment (antiagainst) [reply iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp:222]: I see your points now. Yes it's running after FoldSubViewOps so that we have all intermediate subviews removed. Load/store take the original subspan ops as input, which has all the dynamic dim SSA values associated with them. So this works. Do you see this pattern changing "soonish"? What's coming that will break it?

**Comment by antiagainst** [comment]
@MaheshRavishankar: Previously I didn't check integration tests. It was failing and revealed that we need some more patterns. So I put up another commit 78aac38 to address that. It's non trivial change; so PTAL.

Also @stellaraccident to confirm the patterns in 78aac38 for folding `tensor.dim`/`memref.dim` over shape carrying ops. I slotted it in the shape dialect transformations. That's the place I found most natural. But let me know if there is a better place. (Or maybe there are existing patterns!)

**Review by MaheshRavishankar — APPROVED** [review]
